test/testpushnotifications.cpp

```diff
@@ -0,0 +1,221 @@
+    QString _user;
+    QString _password;
+};
+
+QSharedPointer<OCC::Account> createAccount()
+{
+    auto account = OCC::Account::create();
+
+    QStringList typeList;
+    typeList.append("files");
+
+    QString websocketUrl("ws://localhost:12345");
+
+    QVariantMap endpointsMap;
+    endpointsMap["websocket"] = websocketUrl;
+
+    QVariantMap notifyPushMap;
+    notifyPushMap["type"] = typeList;
+    notifyPushMap["endpoints"] = endpointsMap;
+
+    QVariantMap capabilitiesMap;
+    capabilitiesMap["notify_push"] = notifyPushMap;
+
+    account->setCapabilities(capabilitiesMap);
+
+    return account;
+}
+
+QSharedPointer<OCC::PushNotifications> createPushNotifications(OCC::Account *account)
+{
+    return QSharedPointer<OCC::PushNotifications>(new OCC::PushNotifications(account));
+}
+
+class TestPushNotifications : public QObject
+{
+    Q_OBJECT
+
+private slots:
+    void testSetup_correctCredentials_authenticatedOnWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+
+        QEventLoop loop;
+        QTimer::singleShot(5000, [&]() {
+            loop.quit();
+            QFAIL("No messages arrived");
+        });
+
+        bool userSended = false;
+        MockWebSocketServer mockServer;
+        connect(&mockServer, &MockWebSocketServer::processTextMessage, [&](QWebSocket *socket, QString message) {
+            if (message == user) {
+                userSended = true;
+                return;
+            } else if (message == password) {
+                QCOMPARE(userSended, true);
+                loop.quit();
+                return;
+            }
+            QFAIL("Unexpected message");
+        });
+
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+        pushNotifications->setup();
+
+        loop.exec();
+    }
+
+    void testOnWebSocketTextMessageReceived_notifyFileMessage_emitFilesChanged()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+
+        connect(&mockServer, &MockWebSocketServer::processTextMessage, [&](QWebSocket *sender, QString message) {
+            if (message == password) {
+                sender->sendTextMessage("notify_file");
+                return;
+            }
+        });
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+
+        QEventLoop loop;
+        QTimer::singleShot(5000, [&]() {
+            loop.quit();
+            QFAIL("No files changed event emitted");
+        });
+
+        connect(pushNotifications.data(), &OCC::PushNotifications::filesChanged, [&](OCC::Account * /*account*/) {
```

Comment:
Use QSignalSpy and it's wait method for something like this. You would create the QSignalSpy here and call wait on it instead of your `loop.exec()` call. This should get the QEventLoop to completely go away in this test.